common/workunit/workunit.cpp

```diff
@@ -5462,9 +5490,70 @@ class CDaliWorkUnitFactory : public CWorkUnitFactory, implements IDaliClientShut
+
+                bool hasDynamicAbortingState = false;
+                Owned<IPropertyTree> lWWUPTreeRoot = createPTree("lWWUs");
+                IPropertyTree* wuAbortsPTree = conWorkUnitAborts->queryRoot();
+                ForEach(*originalIter)
+                {
+                    IPropertyTree& cur = originalIter->query();
+
+                    bool dynamicAbortingState = false;
+                    const char* wuid = cur.queryName();
+                    WUState state = getState(&cur, wuAbortsPTree, dynamicAbortingState);
+                    if (!dynamicAbortingState)
+                        lWWUPTreeRoot->addPropTree(wuid, LINK(&cur));
+                    else
+                    {
+                        Owned<IPropertyTree> lWWUPTree = createLightWeitghtWorkunitPTree(&cur, state);
+                        lWWUPTreeRoot->addPropTree(wuid, lWWUPTree.getClear());
+                        hasDynamicAbortingState = true;
+                    }
+                }
+                return hasDynamicAbortingState ? lWWUPTreeRoot->getElements("*") : nullptr;
+            }
+            IPropertyTree* createLightWeitghtWorkunitPTree(IPropertyTree* originalWUTree, WUState& state)
+            {
+                Owned<IPropertyTree> lWWUPTree = createPTree();
+                lWWUPTree->setProp("@jobName", originalWUTree->queryProp("@jobName"));
+                lWWUPTree->setProp("@submitID", originalWUTree->queryProp("@submitID"));
+                lWWUPTree->setProp("@clusterName", originalWUTree->queryProp("@clusterName"));
+                lWWUPTree->setProp("@timeScheduled", originalWUTree->queryProp("@timeScheduled"));
+                lWWUPTree->setProp("@scope", originalWUTree->queryProp("@scope"));
+                lWWUPTree->setProp("Action", originalWUTree->queryProp("Action"));
+                lWWUPTree->setProp("@totalThorTime", originalWUTree->queryProp("@totalThorTime"));
+                lWWUPTree->setProp("@priorityClass", originalWUTree->queryProp("@priorityClass"));
+                lWWUPTree->setPropBool("@protected", originalWUTree->getPropBool("@protected"));
+                setEnum(lWWUPTree, "State", state, states);
+
+                Owned<IPropertyTree> application = originalWUTree->getBranch("Application");
+                if (application)
+                    lWWUPTree->addPropTree("Application", application.getClear());
+                return lWWUPTree.getClear();
+            }
+            WUState getState(IPropertyTree* originalWUTree, IPropertyTree* wuAbortsPTree, bool& dynamicAbortingState)
+            {
+                //In the CLocalWorkUnit::getState(), the checkAgentRunning() is used to get the latest
+                //state for one WU using Dali session manager. But, this getState() may give an
+                //out-of-date state in some unusual circumstances because no effective way can be used
+                //to get the latest state using Dali session manager.
```

Comment:
I'd slightly reword as:

> NB: In CLocalWorkUnit::getState() it also uses checkAgentRunning() to verify the sessionId that the workunit is associated with is still alive. In this implementation the cost of doing that per workunit would be prohibitively expensive, so it's possible in the rare cases when the state is stale because the sessionId is dead, the state from this workunit paging code will be out of date.

Reply:
Will move this comment inside the createWorkunitPTreeIterator() since the getState() here is not needed anymore.